### PR TITLE
886 update derive disposition reason

### LINKED ISSUE #886: Update `derive_disposition_reason()`
`derive_disposition_reason()` should be updated such that there is an option for not populating DSREASP for all values, e.g., only for DSDECOD = 'OTHER'.

### PR DESCRIPTION
**Pull request by wqrena**
close #886.

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

### CODE REVIEW

File: vignettes/adsl.Rmd [hunk 1]
```diff
@@ -271,7 +271,7 @@ This call would return the input dataset with the column `DCSREAS` and `DCSREASP
 By default, the function will map 
 
 - `DCSREAS` as `DSDECOD` if `DSDECOD` is not `"COMPLETED"` or `NA`, `NA` otherwise
-- `DCSREASP` as `DSTERM` if `DSDECOD` is not `"COMPLETED"` or `NA`, `NA` otherwise
+- `DCSREASP` as `DSTERM` if `DSDECOD` is not `"COMPLETED"` or `NA`, and `DSDECOD` is equal to "OTHER",`NA` otherwise
```
Comment (bundfussr) [inline vignettes/adsl.Rmd:274]: A space is missing after the second comma.

I would remove the first comma.

Please use backticks for `"OTHER"`.

File: R/derive_vars_disposition_reason.R [hunk 2]
```diff
@@ -261,15 +265,18 @@ format_reason_default <- function(reason, reason_spe = NULL) {
...
-#'   out <- if ( is.null(reason_spe) ) reason else reason_spe
+#'   out <- if ( is.null(reason_spe) ) reason
+#'     else{
+#'     if_else (reason == "OTHER", reason_spe, NA_character_)
+#'     }
 #'   if_else ( reason != "COMPLETED" & !is.na(reason), out, NA_character_)
 #' }
 #' format_reason_default(DSDECOD) returns `DSDECOD` when `DSDECOD` is not 'COMPLETED' nor NA.
-#' format_reason_default(DSDECOD, DSTERM) returns `DSTERM` when `DSDECOD` is not 'COMPLETED' nor NA.
+#' format_reason_default(DSDECOD, DSTERM) returns `DSTERM` when `DSDECOD` is not 'COMPLETED' nor NA, and `DSDECOD` is equal to "OTHER".
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:279]: This is hard to read in the rendered documentation. We could just link to the documentation of `format_reason_default()`, e.g., something like
```
Default: `format_reason_default`, see [`format_reason_default()`] for details.
```

The "Details" section of `format_reason_default()` needs to be updated.

File: R/derive_vars_disposition_reason.R [hunk 3]
```diff
@@ -261,15 +265,18 @@ format_reason_default <- function(reason, reason_spe = NULL) {
 #' `reason_var` value (e.g. `if_else ( reason != "COMPLETED" & !is.na(reason), reason_spe,
 #' NA_character_)`).
 #' `DCSREASP = format_reason(DSDECOD, DSTERM)` returns DCSREASP = `DSTERM` when `DSDECOD` is not
-#' 'COMPLETED' nor NA.
+#' 'COMPLETED' nor NA, and `DSDECOD` is equal to "OTHER".
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:268]: Please use backticks for `"OTHER"`.

Should the comma be removed?

File: NEWS.md [hunk 4]
```diff
@@ -37,6 +37,7 @@ from a BDS dataset, e.g. ADLB. (#943)
 ## Breaking Changes
 
 - `derive_var_lstalvdt()` has been deprecated in favor of `derive_var_extreme_dt()` (#753).
+- `derive_vars_disposition_reason()` now is updated such that the default is populating DCSREASP only when DSDECOD is equal to "OTHER", which is consistent with ADaMIG_v1.3 (#886).
```
Comment (bundfussr) [inline NEWS.md:40]: Please use backticks for `DCREASP`, `DSDECOD`, and `"OTHER"`.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_vars_disposition_reason.R (hunk 3 above)
Comment (wqrena) [reply R/derive_vars_disposition_reason.R:268]: I also updated other parts similar to 'OTHER' that does not have backticks (eg 'COMPLETED'). Please review again to see if it is ok.

**Comment by wqrena** [comment]
@bundfussr Hi Stefan, thanks for the review. All the suggested part has been updated. I also updated other code parts in the header of R file that does not have backticks. Please check to see if they are ok. thanks!

**Comment by bms63** [comment]
can you run `devtools::document` to see if R-CMD latest check resolves itself.  It is falling over and I am not quite sure why

**Comment by wqrena** [comment]
Hi Ben,
I ran both devtools::document() and devtools::check() locally. They seem fine to me.

**Comment by bms63** [comment]
Can you add hms to this file for importfrom lubridate and rerun devtools::document() and push up 
image

**Comment by wqrena** [comment]
Hi Ben, done as suggested .

**Comment by bms63** [comment]
fingers crossed

**Comment by wqrena** [comment]
Hi Ben,
R-CMD-check runs fine on github a couple of days ago. The recent branch updates only touched a few documentation such as  help manual, vignette and NEWs. Not sure how NAMESPACE in importFrom get impacted now on the github R-CMD-check whereas devtools::check() runs fine locally. 
I also noticed R-CMD-check failed for other PR on github with similar reason (#776). Maybe some other reason leading to the error ?

File: R/derive_vars_disposition_reason.R [hunk 5]
```diff
@@ -198,7 +197,11 @@ derive_disposition_reason <- function(dataset,
...
 #'
 format_reason_default <- function(reason, reason_spe = NULL) {
-  out <- if (is.null(reason_spe)) reason else reason_spe
+  out <- if (is.null(reason_spe)) reason
+  else{
+    if_else (reason == "OTHER", reason_spe, NA_character_)
+  }
+
   if_else(reason != "COMPLETED" & !is.na(reason), out, NA_character_)
 }
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:206]: Could we simplify this to:
```
if (is.null(reason_spe)) {
  if_else(reason != "COMPLETED" & !is.na(reason), reason, NA_character_)
} else {
  if_else (reason == "OTHER", reason_spe, NA_character_)
}

File: R/derive_vars_disposition_reason.R [hunk 6]
```diff
@@ -170,9 +170,8 @@ derive_disposition_reason <- function(dataset,
 #' if required (e.g. `DSTERM`).
 #'
 #' @details
-#' `format_reason_default(DSDECOD)` returns `DSDECOD` when `DSDECOD` is not
-#' 'COMPLETED' nor `NA`. `format_reason_default(DSDECOD, DSTERM)` returns
-#' `DSTERM` when `DSDECOD` is not 'COMPLETED' nor `NA`.
+#' `format_reason_default(DSDECOD)` returns `DSDECOD` when `DSDECOD` is not `'COMPLETED'` nor `NA`.
+#' \cr`format_reason_default(DSDECOD, DSTERM)` returns `DSTERM` when `DSDECOD` is not `'COMPLETED'` nor `NA` and `DSDECOD` is equal to `'OTHER'`.
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:174]: Could we simplify it by removing "`DSDECOD` is not `'COMPLETED'` nor `NA` and "?

File: R/derive_vars_disposition_reason.R [hunk 7]
```diff
@@ -246,30 +249,24 @@ format_reason_default <- function(reason, reason_spe = NULL) {
...
-#' `new_var_spe` will be derived as `new_var_spe` = `format_reason(reason_var, reason_var_spe)`,
+#' `new_var` will be derived as `new_var = format_reason(reason_var)`,
+#' `new_var_spe` will be derived as `new_var_spe = format_reason(reason_var, reason_var_spe)`.
 #' Typically, the content of the function would return `reason_var_spe` or `NA` depending on the
 #' `reason_var` value (e.g. `if_else ( reason != "COMPLETED" & !is.na(reason), reason_spe,
 #' NA_character_)`).
-#' `DCSREASP = format_reason(DSDECOD, DSTERM)` returns DCSREASP = `DSTERM` when `DSDECOD` is not
-#' 'COMPLETED' nor NA.
+#' `DCSREASP = format_reason(DSDECOD, DSTERM)` returns `DCSREASP = DSTERM` when `DSDECOD` is not
+#' `'COMPLETED'` nor `NA` and `DSDECOD` is equal to `'OTHER'`.
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:267]: As above

File: R/derive_vars_disposition_reason.R [hunk 8]
```diff
@@ -321,7 +318,10 @@ format_reason_default <- function(reason, reason_spe = NULL) {
...
 #' # Derive DCSREAS and DCSREASP using a study-specific format
 #' format_dcsreas <- function(x, y = NULL) {
-#'   out <- if (is.null(y)) x else y
+#'   out <- if (is.null(y)) x
+#'      else {
+#'      if_else(x == "OTHER", y, NA_character_)
+#'      }
 #'   case_when(
 #'     !(x %in% c("COMPLETED", "SCREEN FAILURE")) & !is.na(x) ~ out,
 #'     TRUE ~ NA_character_
```
Comment (bundfussr) [inline R/derive_vars_disposition_reason.R:327]: We could simplify this to:
```
if (is.null(reason_spe)) {
  if_else(!(x %in% c("COMPLETED", "SCREEN FAILURE")) & !is.na(x) , reason, NA_character_)
} else {
  if_else (reason == "OTHER", reason_spe, NA_character_)
}
```

File: tests/testthat/test-derive_vars_disposition_reason.R [hunk 9]
```diff
@@ -50,19 +49,49 @@ test_that("Derive DCSREAS using default mapping", {
...
   format_dctreas <- function(x, y = NULL) {
-    out <- if (is.null(y)) x else y
+    out <- if (is.null(y)) x
+    else {
+      if_else(x == "OTHER", y, NA_character_)
+    }
     case_when(
       x %notin% c("COMPLETED", "SCREEN FAILURE") & !is.na(x) ~ out,
       TRUE ~ NA_character_
     )
```
Comment (bundfussr) [inline tests/testthat/test-derive_vars_disposition_reason.R:88]: See above

File: vignettes/adsl.Rmd [hunk 10]
```diff
@@ -271,7 +271,7 @@ This call would return the input dataset with the column `DCSREAS` and `DCSREASP
 By default, the function will map 
 
 - `DCSREAS` as `DSDECOD` if `DSDECOD` is not `"COMPLETED"` or `NA`, `NA` otherwise
-- `DCSREASP` as `DSTERM` if `DSDECOD` is not `"COMPLETED"` or `NA`, `NA` otherwise
+- `DCSREASP` as `DSTERM` if `DSDECOD` is not `"COMPLETED"` or `NA` and `DSDECOD` is equal to `OTHER`, `NA` otherwise
```
Comment (bundfussr) [inline vignettes/adsl.Rmd:274]: See above

File: vignettes/adsl.Rmd [hunk 11]
```diff
@@ -282,7 +282,10 @@ Example function `format_dcsreas()`:
...
 ```{r eval=TRUE}
 format_dcsreas <- function(dsdecod, dsterm = NULL) {
-  out <- if (is.null(dsterm)) dsdecod else dsterm
+  out <- if (is.null(dsterm)) dsdecod
+  else{
+    if_else (dsdecod == "OTHER", dsterm, NA_character_)
+  }
   case_when(
     !dsdecod %in% c("COMPLETED", "SCREEN FAILURE") & !is.na(dsdecod) ~ out,
     TRUE ~ NA_character_
```
Comment (bundfussr) [inline vignettes/adsl.Rmd:291]: See above

**Review by bundfussr — CHANGES_REQUESTED** [review]
Sorry for requesting more changes. I just realized that the new descriptions and code could be simplified.

File: R/derive_vars_disposition_reason.R (hunk 5 above)
Comment (wqrena) [reply R/derive_vars_disposition_reason.R:206]: Updated

File: R/derive_vars_disposition_reason.R (hunk 6 above)
Comment (wqrena) [reply R/derive_vars_disposition_reason.R:174]: Updated.

File: R/derive_vars_disposition_reason.R (hunk 7 above)
Comment (wqrena) [reply R/derive_vars_disposition_reason.R:267]: Updated

File: vignettes/adsl.Rmd (hunk 10 above)
Comment (wqrena) [reply vignettes/adsl.Rmd:274]: Updated

**Comment by wqrena** [comment]
Thank you, Stefan and Ben!
Update is completed based on the suggestion.

**Review by bundfussr — APPROVED** [review]

**Comment by bundfussr** [comment]
@bms63 , could you merge this PR using admin privileges?

**Comment by bms63** [comment]
Look at all those green checks!

**Comment by bundfussr** [comment]
@bms63 , that's great but we still have the issue that there are two R 3.6 checks and the required one does not run. This happens even in very recent PRs like https://github.com/pharmaverse/admiral/pull/1098. Any ideas?

**Comment by bms63** [comment]
It is on my list! :)  Hopefully an easy fix!

**Pull request merged**